[DRAFT] Interp mode testing

davidwrighton opened the pull request:
Adjust interpmode to automatically work in a few more ways, so that we can run a larger test run with InterpMode=3

davidwrighton committed:
InterpMode testing

davidwrighton committed:
InterpMode=3 change for Bash targets

davidwrighton commented:
/azp run runtime-interpreter

davidwrighton commented:
/azp run runtime-interpreter

jkotas commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -1195,6 +1195,15 @@ void EEJitManager::SetCpuInfo()
+    bool allowHWIntrinsic = true;
+
+#if defined(FEATURE_INTERPRETER)
+    if (maxVectorTBitWidth != 128 && CLRConfig::GetConfigValue(CLRConfig::EXTERNAL_InterpMode) >= 2)
```
Can this be `GetConfigValue(CLRConfig::EXTERNAL_InterpMode) == 3`?

davidwrighton committed:
[clr-interp] add support for jmp instruction

davidwrighton committed:
Make it build in retail

davidwrighton committed:
[clr-interp] Fix stack alignment of 16 byte or greater aligned structures in the interpreter
- Arguments are up to 16 byte aligned
- Handle this in the call stub generator, and call emission. (We already did this in the argument input processing in the interpreter

We need to align these at 16byte boundaries as the various ABIs tend to have a hard requirement around 16 byte alignment, when they don't have such a requirement around 32 or 64 byte alignment.

davidwrighton committed:
Merge branch 'main' of https://github.com/dotnet/runtime into InterpModeTesting

davidwrighton committed:
Enable running all the interp modes

davidwrighton committed:
Merge branch 'main' of https://github.com/dotnet/runtime into InterpModeTesting

davidwrighton committed:
[clr-interp] Fix issue where switch instructions did not properly handle dispatch to blocks which already had defined their incoming stack slots

davidwrighton committed:
[clr-interp] ret instructions are similar to unconditional branches in that they do not link to the next block

davidwrighton committed:
Revert "[clr-interp] add support for jmp instruction"

This reverts commit 004aa94201af7d96fbc699ba3307622cec920de0.

davidwrighton committed:
[clr-interp] Add a scheme for adding peephole optimizations for known sequences of IL opcodes
- This is needed to ensure that some of our SIMD tests finish in a vaguely reasonable timeframe
- The peeps implmented here make that almost possible, although the behavior in CI hasn't yet been verified
- Peeps implemented
  - stloc/ldloc  ... This is a minor improvement, but will be needed to handle a future optimization around allowing the il stack to have constant values
  - box/unbox.any - Removes unnecessary boxing/unboxing
  - typeof(T)==typeof(Y) - This allows us to handle the type testing specialization behavior that is used heavily in the BCL
  - typeof(T).IsValueType - Used in the Unsafe.BitCast function

Not yet implemented is giving the interpreter stack a concept of constant values, so that we can optimize brtrue/brfalse and friends into unconditional branches. With this set of changes the if (typeof(T)==typeof(int)) { ... } else if (typeof(T)==typeof(float)) pattern is *much* faster than before, but the not taken paths are still fully generated interpreter code, and there are still many branches.

davidwrighton committed:
Tweak what interpmode stuff to run
- Run all the scenarios that do anything with the interpreter
- Don't run OSX since that pool seems busy these days, and this is a big run

davidwrighton committed:
Merge branch 'main' of https://github.com/dotnet/runtime into InterpModeTesting

davidwrighton commented:
/azp run runtime-interpreter

davidwrighton committed:
Merge branch 'main' into InterpModeTesting

davidwrighton committed:
Allow InterpMode to be set in the testenv

davidwrighton committed:
Merge branch 'InterpModeTesting' of https://github.com/davidwrighton/runtime into InterpModeTesting

davidwrighton commented:
/azp run runtime-interpreter

janvorli committed:
Disable CoreCLR tests crashing merged test suites fully interpreted

This change disables all coreclr tests that cause the merged coreclr
test suites to crash and so preventing us from getting any test pass /
fail information from those. This is with running the tests with
InterpMode=3 (fully interpreted)

This was put together on Windows x64 checked build, we may need to add
more for other targets in a separate PR.

janvorli commented on eng/pipelines/coreclr/interpreter.yml:
```diff
@@ -73,8 +73,6 @@ extends:
           - linux_arm64
           - windows_x64
           - windows_arm64
-          - osx_x64
```
Why have you removed the osx_arm64? It is actually the most important configuration.

davidwrighton replied:
That was just temporary. I was trying to reduce the load on the servers as I did this big runs to find out if the CI could be coerced into getting good data.

janvorli committed:
PR feedback - update issue links

davidwrighton committed:
Merge branch 'main' of https://github.com/dotnet/runtime into InterpModeTesting

davidwrighton committed:
Fix issues

davidwrighton commented:
/azp run runtime-interpreter
